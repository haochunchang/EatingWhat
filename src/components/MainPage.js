import React, { Component } from "react";
import loadable from "@loadable/component";
import "./MainPage.css";

import Header from "./Header";
import Scroll from "./Scroll";
import ErrorBoundary from "../containers/ErrorBoundary";

const CardList = loadable(() => import("./CardList"));
const SearchBox = loadable(() => import("./SearchBox"));
const StarchFoodButton = loadable(() => import("./StarchFoodButton"));
const ProteinFoodButton = loadable(() => import("./ProteinFoodButton"));

class MainPage extends Component {
  constructor() {
    super();
    this.state = {
      starch_index: -1,
      protein_index: -1,
    };
  }

  componentDidMount() {
    this.props.onRequestFoods();
  }

  onClickedStarch() {
    const srandom = Math.floor(Math.random() * this.props.starchyFoods.length);
    this.setState({
      starch_index: srandom,
      searchfield: "",
    });
  }

  onClickedProtein() {
    const prandom = Math.floor(Math.random() * this.props.proteinFoods.length);
    this.setState({
      protein_index: prandom,
      searchfield: "",
    });
  }

  filterFoods = (index, foods) => {
    var filteredFoods = [];
    if (index !== -1 && !this.props.searchfield.length) {
      filteredFoods = [foods[index]];
    } else if (foods.length > 0) {
      filteredFoods = foods.filter((food) => {
        return food.name
          .toLowerCase()
          .includes(this.props.searchfield.toLowerCase());
      });
    }
    return filteredFoods;
  };

  render() {
    const { starch_index, protein_index } = this.state;
    const {
      starchyFoods,
      proteinFoods,
      onSearchChange,
      isPending,
    } = this.props;
    var StarchyFoods = this.filterFoods(starch_index, starchyFoods);
    var ProteinFoods = this.filterFoods(protein_index, proteinFoods);

    return isPending ? (
      <div className="App">
        <header className="header">
          <h1 className="f1">Loading...</h1>
        </header>
      </div>
    ) : (
      <div className="App">
        <Header />
        <div className="row">
          <SearchBox searchChange={onSearchChange} />

          <div className="column">
            <StarchFoodButton onClicked={this.onClickedStarch.bind(this)} />
            <Scroll>
              <ErrorBoundary>
                <CardList foods={StarchyFoods} />
              </ErrorBoundary>
            </Scroll>
          </div>

          <div className="column">
            <ProteinFoodButton onClicked={this.onClickedProtein.bind(this)} />
            <Scroll>
              <ErrorBoundary>
                <CardList foods={ProteinFoods} />
              </ErrorBoundary>
            </Scroll>
          </div>
        </div>
      </div>
    );
  }
}

export default MainPage;
