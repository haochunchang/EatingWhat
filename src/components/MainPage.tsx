import React, { Component } from "react";
import loadable from "@loadable/component";
import "./MainPage.css";

import Header from "./Header";
import Scroll from "./Scroll";
import ErrorBoundary from "../containers/ErrorBoundary";

const CardList = loadable(() => import("./CardList"));
const SearchBox = loadable(() => import("./SearchBox"));
const StarchFoodButton = loadable(() => import("./StarchFoodButton"));
const ProteinFoodButton = loadable(() => import("./ProteinFoodButton"));

export interface Food {
  id: number;
  name: string;
  category: string;
  url: string;
}

export interface AppProps {
  starchyFoods: Array<Food>;
  proteinFoods: Array<Food>;
  searchfield: string;
  isPending: boolean;
  error: string;

  onSearchChange(event: React.SyntheticEvent<HTMLInputElement>): void;
  onRequestFoods(): void;
}

export interface AppState {
  starch_index: number;
  protein_index: number;
  searchfield?: string;

  searchFoods: {
    searchfield: string;
  };
  requestFoods: {
    foods: Array<Food>;
    isPending: boolean;
    error: string;
  };
}

class MainPage extends Component<AppProps, AppState> {
  constructor(props: AppProps) {
    super(props);
    this.state = {
      starch_index: -1,
      protein_index: -1,
      searchFoods: {
        searchfield: "",
      },
      requestFoods: {
        foods: [],
        isPending: false,
        error: "",
      },
    };
  }

  componentDidMount(): void {
    this.props.onRequestFoods();
  }

  onClickedStarch(): void {
    const srandom = Math.floor(Math.random() * this.props.starchyFoods.length);
    this.setState({
      starch_index: srandom,
      searchfield: "",
    });
  }

  onClickedProtein(): void {
    const prandom = Math.floor(Math.random() * this.props.proteinFoods.length);
    this.setState({
      protein_index: prandom,
      searchfield: "",
    });
  }

  filterFoods = (index: number, foods: Array<Food>): Array<Food> => {
    var filteredFoods: Array<Food> = [];
    if (index !== -1 && !this.props.searchfield.length) {
      filteredFoods = [foods[index]];
    } else if (foods.length > 0) {
      filteredFoods = foods.filter((food) => {
        return food.name
          .toLowerCase()
          .includes(this.props.searchfield.toLowerCase());
      });
    }
    return filteredFoods;
  };

  render(): JSX.Element {
    console.log(this.state);
    const { starch_index, protein_index } = this.state;
    const {
      starchyFoods,
      proteinFoods,
      isPending,
      onSearchChange,
    } = this.props;
    var StarchyFoods = this.filterFoods(starch_index, starchyFoods);
    var ProteinFoods = this.filterFoods(protein_index, proteinFoods);

    return isPending ? (
      <div className="App">
        <header className="header">
          <h1 className="f1">Loading...</h1>
        </header>
      </div>
    ) : (
      <div className="App">
        <Header />
        <div className="row">
          <SearchBox searchChange={onSearchChange} />

          <div className="column">
            <StarchFoodButton onClicked={this.onClickedStarch.bind(this)} />
            <Scroll>
              <ErrorBoundary>
                <CardList foods={StarchyFoods} />
              </ErrorBoundary>
            </Scroll>
          </div>

          <div className="column">
            <ProteinFoodButton onClicked={this.onClickedProtein.bind(this)} />
            <Scroll>
              <ErrorBoundary>
                <CardList foods={ProteinFoods} />
              </ErrorBoundary>
            </Scroll>
          </div>
        </div>
      </div>
    );
  }
}

export default MainPage;
