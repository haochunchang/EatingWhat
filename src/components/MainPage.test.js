import React from "react";
import { shallow } from "enzyme";
import MainPage from "./MainPage";

describe("MainPage testing", () => {
  let wrapper;
  let mainpage;
  let mockProps;

  beforeEach(() => {
    mockProps = {
      onRequestFoods: jest.fn(),
      onSearchChange: jest.fn(),
      starchyFoods: [
        {
          id: 1,
          name: "Soba",
          category: "main food",
        },
      ],
      proteinFoods: [
        {
          id: 5,
          name: "Steak",
          category: "meat",
        },
      ],
      searchfield: "",
      isPending: false,
    };
    wrapper = shallow(<MainPage {...mockProps} />);
    mainpage = wrapper.instance();
  });

  it("renders MainPage correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("test componentDidMount", async () => {
    expect.assertions(1);
    await mainpage.componentDidMount();
    expect(mainpage.props.onRequestFoods).toHaveBeenCalledTimes(2);
  });

  it("filter foods by searchfield", () => {
    expect(mainpage.filterFoods(-1, [])).toEqual([]);
    expect(mainpage.filterFoods(-1, mockProps.proteinFoods)).toEqual(
      mockProps.proteinFoods
    );

    mockProps.searchfield = "st";
    const mainpage_with_searchfield = shallow(
      <MainPage {...mockProps} />
    ).instance();
    expect(
      mainpage_with_searchfield.filterFoods(-1, mockProps.starchyFoods)
    ).toEqual([]);
  });

  it("tests onClick event on food button", () => {
    mainpage.onClickedStarch();
    mainpage.onClickedProtein();
    expect(mainpage.state.starch_index).toBeGreaterThan(-1);
    expect(mainpage.state.protein_index).toBeGreaterThan(-1);
  });
});
