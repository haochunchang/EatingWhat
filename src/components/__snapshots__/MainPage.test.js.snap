// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage testing renders MainPage correctly 1`] = `ShallowWrapper {}`;
